### PR TITLE
add checkstyle (no tabs, left curly) and usage docs #5075

### PR DESCRIPTION
**Pull request by pdurbin**
connects to #5075

### CODE REVIEW

File: checkstyle.xml [hunk 1]
```diff
@@ -0,0 +1,211 @@
+<?xml version="1.0"?>
+<!DOCTYPE module PUBLIC
+          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
+          "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
+
+<!--
+
+  Checkstyle configuration that checks the sun coding conventions from:
```
Comment (pameyer) [inline checkstyle.xml]: Code says "sun conventions"; docs say "default Netbeans formatting settings".  Are these different, and if so which one is authoritative?

File: doc/sphinx-guides/source/developers/coding-style.rst [hunk 2]
```diff
@@ -59,10 +59,20 @@ Format Code You Changed with Netbeans
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-We would like to someday automate the detection and possibly correction of code that hasn't been formatted using our house style (the default Netbeans style). We've heard that https://maven.apache.org/plugins/maven-checkstyle-plugin/ can do this but we would be happy to see a pull request in this area, especially if it also hooks up to our builds at https://travis-ci.org/IQSS/dataverse .
+The easiest way to adopt Dataverse coding style is to use Netbeans as your IDE, avoid change the default Netbeans formatting settings, and only reformat code you've changed, as described above.
+
+If you do not use Netbeans, you are encouraged to check the formatting of your code using Checkstyle.
+
+To check the entire project:
+
+``mvn checkstyle:checkstyle``
```
Comment (pameyer) [inline doc/sphinx-guides/source/developers/coding-style.rst]: If possible, html checkstyle reports might be more readable than xml reports.

File: doc/sphinx-guides/source/developers/coding-style.rst [hunk 3]
```diff
@@ -59,10 +59,20 @@ Format Code You Changed with Netbeans
...
+
+To check a single file:
+
+``mvn checkstyle:checkstyle -Dcheckstyle.includes=**\/SystemConfig*.java``
```
Comment (pameyer) [inline doc/sphinx-guides/source/developers/coding-style.rst]: Does checkstyle support reformatting (either individual files, or ideally "only reformat code you've changed")?

File: pom.xml [hunk 4]
```diff
@@ -641,6 +641,16 @@
                     <argLine>-Duser.timezone=${project.timezone} -Dfile.encoding=${project.build.sourceEncoding} -Duser.language=${project.language} -Duser.region=${project.region}</argLine>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-checkstyle-plugin</artifactId>
+                <version>3.0.0</version>
+                <configuration>
+                    <configLocation>checkstyle.xml</configLocation>
```
Comment (pameyer) [inline pom.xml]: Is the convention for `checkstyle.xml` to live at the top-level, or should it live somewhere else (`conf/`? )?

File: doc/sphinx-guides/source/developers/coding-style.rst (hunk 2 above)
Comment (pdurbin) [reply doc/sphinx-guides/source/developers/coding-style.rst]: It's a great idea to have HTML reports from Checkstyle but from a quick look at https://stackoverflow.com/questions/6342211/how-to-generate-checkstyle-reports we might be able to use `mvn site` to create the HTML for. I played with `mvn site` 6 or so years ago at https://github.com/dvn/dvn-mavensitepoc

File: doc/sphinx-guides/source/developers/coding-style.rst (hunk 3 above)
Comment (pdurbin) [reply doc/sphinx-guides/source/developers/coding-style.rst]: I don't know.

**Comment by pdurbin** [comment]
@qqmyers thanks for all of your thoughts on this. In 6ca8617 I specified 4 spaces for Java. For Bash I said 2 spaces in pull request #5063.

File: checkstyle.xml (hunk 1 above)
Comment (pdurbin) [reply checkstyle.xml]: Removed in 80334fc

File: doc/sphinx-guides/source/developers/coding-style.rst (hunk 3 above)
Comment (poikilotherm) [reply doc/sphinx-guides/source/developers/coding-style.rst]: Nope. It will only validate things, not fix it for you.

File: pom.xml (hunk 4 above)
Comment (poikilotherm) [reply pom.xml]: That depends on your personal opinion. From the Java projects I know, people just place it at the same level as the `pom.xml`.

Comment (pdurbin) [reply pom.xml]: I like it when I find certain files at the top level, including:

- README.md
- LICENSE.md
- CONTRIBUTING.md
- Dockerfile
- Vagrantfle

and for Java projects, seeing checkstyle.xml at the top level seems nice. 😄

File: doc/sphinx-guides/source/developers/coding-style.rst (hunk 3 above)
Comment (pdurbin) [reply doc/sphinx-guides/source/developers/coding-style.rst]: If we switch to Go, we can use gofmt, which will reformat as well. 😄 See https://groups.google.com/d/msg/dataverse-dev/y2Jpk3szTf8/rckKmP6-BgAJ

File: checkstyle.xml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            <property name="ignoreNamePattern" value="^[A-Z_\-]+$" />
+        </module>
+        <module name="JavadocStyle">
+            <property name="checkFirstSentence" value="false" />
+        </module>
+        -->
+
+        <!-- Checks for Naming Conventions.                  -->
+        <!-- See http://checkstyle.sf.net/config_naming.html -->
+        <!--
```
Comment (michbarsinai) [inline checkstyle.xml]: Should we enable these? I think we're in pretty good shape in this area, but it's good to have this enforces, especially when people with non-Java (=no camelCase) background contribute.

**Pull request merged**